tests/test_core/test_future.py

```diff
@@ -0,0 +1,74 @@
+import os
+import sys
+import logging
+
+from collections import OrderedDict
+
+sys.path.append(os.path.join(os.path.dirname(__file__), ".."))
+import setup_malcolm_paths
+
+import unittest
+from mock import MagicMock, call, ANY
+
+#module imports
+from malcolm.compat import queue
+from malcolm.core.task import Task
+from malcolm.core.future import Future, TimeoutError, RemoteError
+from malcolm.core.response import Response
+
+import logging
+logging.basicConfig(level=logging.DEBUG)
+
+class TestFuture(unittest.TestCase):
+
+    def setUp(self):
+        self.proc = MagicMock(q=queue.Queue())
+        self.proc.create_queue = MagicMock(return_value=queue.Queue())
+        self.task = Task("testTask", self.proc)
+
+    def test_init(self):
+        f = Future(self.task)
+        self.assertEqual(f._task, self.task)
+
+    def test_set_result(self):
+        f = Future(self.task)
+        f.set_result("testResult")
+        self.assertTrue(f.done())
+        self.assertEqual(f.result(0), "testResult")
+
+    def test_set_exception(self):
+        f = Future(self.task)
+        f.set_exception("test Error")
+        self.assertTrue(f.done())
+        self.assertRaises(RemoteError, f.result, 0)
+        self.assertEqual(f.exception(),'test Error')
+
+    def test_result(self):
+        # timeout due to no response arriving
```

Comment:
Yes - if you call exception on an unfilled future then it does a Task.wait_all(0) on its self and raises Queue.Empty once time-out is reached (in this test time-out is zero)